aws-signer/src/main/scala/http4s/aws/middleware/AwsSigner.scala

```diff
@@ -0,0 +1,136 @@
+package http4s.aws.middleware
+
+import java.util.Date
+
+import cats.data.Kleisli
+import cats.effect.Sync
+import cats.implicits._
+import fs2.Stream
+import org.http4s.client.Client
+import org.http4s.{Header, Request}
+import java.nio.charset.StandardCharsets
+
+import BytesToHex.bytesToHex
+
+object AwsSigner {
+  def apply[F[_]: Sync](id: String, secret: String, zone: String, service: String)(
+      client: Client[F]): Client[F] = {
+    val signer = new AwsSigner(Key(id, secret), zone, service)
+    client.copy(open = Kleisli { req =>
+      signer[F](req, Sync[F].delay(new Date)).flatMap(client.open(_))
+    })
+  }
+
+  case class Key(id: String, secret: String)
+}
+
+/** Implements AWS request signing v4 as an http4s middleware.
+  * Signing described here: https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html
+  */
+class AwsSigner(key: AwsSigner.Key, zone: String, service: String) {
```

Comment:
Do you want this constructor to be public as an alternate way of creating one with a `Key`?  If not, you can whittle down the surface area of the API some more with a private constructor, and then even make `Key` private.